### PR TITLE
Implement Credential-Based Managed Identity Authentication

### LINKED ISSUE #4411: Add support for credential based managed identity
Implement MSI V2 with /credential endpoint, ensuring robust token acquisition, SLC revocation handling, and probe logic for VM/VMSS environments.

**Key Deliverables**
MSI V2 /credential Endpoint Implementation

- [ ] Develop a design document detailing the token acquisition process for MSI V2.
- [ ] Include certificate handling procedures and source detection logic.
- [ ] Ensure clear documentation for SDK developers on integrating the /credential endpoint.
- [ ] Short-Lived Credential (SLC) Revocation Specification
- [ ] Define a structured approach for handling SLC revocation scenarios.
- [ ] Outline the process for acquiring new credentials when an existing one is revoked or expires.
- [ ] Ensure proper handling of revocation-related failures and retry mechanisms.
- [ ] VM/VMSS Credential Endpoint Probe Logic

**Implement and document the probe logic for detecting the MSI V2 /credential endpoint in IMDS.**
- [ ] Address scenarios related to IMDS restarts and expected response handling.
- [ ] Provide guidance on how SDKs should interpret probe results and fallback strategies.

**Success Criteria**
- [ ] Comprehensive documentation and design specifications are available for developers.
- [ ] Clear and actionable steps for integrating MSI V2 /credential endpoint, SLC revocation, and probing logic.
- [ ] Validation of probe logic ensuring seamless integration in VM/VMSS environments.

### PR DESCRIPTION
**Pull request by gladjohn**
Fixes #4411

**Changes proposed in this request**
- This new implementation focuses on supporting Managed Identity authentication using credentials. It caters specifically to scenarios where Managed Identity is utilized using a machine key as a source
- MSAL creates an in-memory certificate using the MI key, uses the certificate to get the credential, passes the credential as an assertion to ESTS-R over MTLS with the same certificate 
- HttpManager changes have been made to support MTLS
- Added NET472, as this is the minimum supported for MI with credential as a source 
- Adds claims and capabilities support for Managed Identity 
- Supports SLC on Windows VMs where a software or a machine key has been provisioned 

Please review [SLC board](https://github.com/orgs/AzureAD/projects/27/views/1) for a full list of work items. 

**Testing**
- unit tests
- need to add integration tests

**Performance impact**
none

**Documentation**
n/a

### CODE REVIEW

**Comment by SimonCropp** [comment]
re "Documentation: n/a". this adds new public apis. how can there be no associated documentation?

File: src/client/Microsoft.Identity.Client/AppConfig/ManagedIdentityApplicationBuilder.cs [hunk 1]
```diff
@@ -100,6 +100,28 @@ public ManagedIdentityApplicationBuilder WithTelemetryClient(params ITelemetryCl
...
+        /// </summary>
+        /// <remarks>
+        /// MSAL will transform these into special claims request. See https://openid.net/specs/openid-connect-core-1_0-final.html#ClaimsParameter for
+        /// details on claim requests. This is an experimental API. The method signature may change in the future 
+        /// without involving a major version upgrade.
+        /// For more details see https://aka.ms/msal-net-claims-request
+        /// </remarks>
+        public ManagedIdentityApplicationBuilder WithClientCapabilities(IEnumerable<string> clientCapabilities)
+        {
+            ValidateUseOfExperimentalFeature("WithClientCapabilities");
```
Comment (SimonCropp) [inline src/client/Microsoft.Identity.Client/AppConfig/ManagedIdentityApplicationBuilder.cs:115]: `"WithClientCapabilities"` is redundant

File: src/client/Microsoft.Identity.Client/Credential/ManagedIdentityCredentialResponse.cs [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        {
+            _requestContext.Logger.Info("[Managed Identity] Fetching new managed identity credential from IMDS endpoint.");
+
+            OAuth2Client client = CreateClientRequest(httpManager);
+
+            CredentialResponse credentialResponse = await client
+                .GetCredentialResponseAsync(_uri, _requestContext)
+                .ConfigureAwait(false);
+
+            return credentialResponse;
```
Comment (SimonCropp) [inline src/client/Microsoft.Identity.Client/Credential/ManagedIdentityCredentialResponse.cs:131]:
```suggestion
            return await client
                .GetCredentialResponseAsync(_uri, _requestContext)
                .ConfigureAwait(false);
```

File: src/client/Microsoft.Identity.Client/Credential/ManagedIdentityCredentialResponse.cs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+                            ""jwk"": {
+                                ""kty"": ""RSA"", 
+                                ""use"": ""sig"",
+                                ""alg"": ""RS256"",
+                                ""kid"": """ + x509Certificate2.Thumbprint + @""",
+                                ""x5c"": [""" + certificateBase64 + @"""]
+                            }
+                        },
+                        ""latch_key"": false    
+                    }";
```
Comment (SimonCropp) [inline src/client/Microsoft.Identity.Client/Credential/ManagedIdentityCredentialResponse.cs:172]:
```suggestion
            return $$"""
                     {
                         "cnf": {
                             "jwk": {
                                 "kty": "RSA",
                                 "use": "sig",
                                 "alg": "RS256",
                                 "kid": "{{x509Certificate2.Thumbprint}}",
                                 "x5c": ["{{certificateBase64}}"]
                             }
                         },
                         "latch_key": false
                     }
                     """;
```

File: src/client/Microsoft.Identity.Client/Http/HttpManager.cs [hunk 4]
```diff
@@ -25,123 +26,41 @@ namespace Microsoft.Identity.Client.Http
...
-            CancellationToken cancellationToken = default)
+            bool doNotThrow,
+            bool retry,
+            X509Certificate2 bindingCertificate,
+            CancellationToken cancellationToken)
         {
+            Exception timeoutException = null;
             HttpResponse response = null;
-            
+            bool isRetriable = false;
```
Comment (SimonCropp) [inline src/client/Microsoft.Identity.Client/Http/HttpManager.cs:62]:
```suggestion
            bool isRetriable;
```

File: src/client/Microsoft.Identity.Client/Http/HttpManager.cs [hunk 5]
```diff
@@ -198,11 +149,39 @@ private static HttpRequestMessage CreateRequestMessage(Uri endpoint, IDictionary
...
+                return msalMtlsHttpClientFactory.GetHttpClient(x509Certificate2);
+            }
+
+            // If the factory is not an IMsalMtlsHttpClientFactory, use it to get a default HttpClient
+            return _httpClientFactory.GetHttpClient();
+        }
+
+        private static HttpRequestMessage CreateRequestMessage(Uri endpoint, IDictionary<string, string> headers)
+        {
+            HttpRequestMessage requestMessage = new HttpRequestMessage { RequestUri = endpoint };
```
Comment (SimonCropp) [inline src/client/Microsoft.Identity.Client/Http/HttpManager.cs:166]:
```suggestion
            var requestMessage = new HttpRequestMessage { RequestUri = endpoint };
```

File: src/client/Microsoft.Identity.Client/Http/HttpManager.cs [hunk 6]
```diff
@@ -273,13 +239,20 @@ protected static async Task<HttpContent> CloneHttpContentAsync(HttpContent httpC
             return clone;
         }
 
-        /// <summary>
-        /// In HttpManager, the retry policy is based on this simple condition.
-        /// Avoid changing this, as it's breaking change.
-        /// </summary>
-        protected virtual bool IsRetryableStatusCode(int statusCode)
+        #region Helpers
```
Comment (SimonCropp) [inline src/client/Microsoft.Identity.Client/Http/HttpManager.cs:242]: IMO the regions dont add enough value

File: src/client/Microsoft.Identity.Client/Instance/Oidc/OidcRetrieverWithCache.cs [hunk 7]
```diff
@@ -39,25 +39,33 @@ internal static class OidcRetrieverWithCache
...
+                    requestContext.ServiceBundle.HttpManager,
+                    mtlsCertificate: null);
+
+                configuration = await client.DiscoverOidcMetadataAsync(oidcMetadataEndpoint, requestContext)
+                    .ConfigureAwait(false);
 
                 s_cache[authority] = configuration;
 
-                requestContext.Logger.Verbose(() => $"[OIDC Discovery] OIDC discovery retrieved metadata from the network for {authority}");
+                requestContext.Logger.Verbose(() => $"[OIDC Discovery] OIDC discovery retrieved metadata from the " +
```
Comment (SimonCropp) [inline src/client/Microsoft.Identity.Client/Instance/Oidc/OidcRetrieverWithCache.cs:52]: why the concat?

File: src/client/Microsoft.Identity.Client/Instance/Oidc/OidcRetrieverWithCache.cs [hunk 8]
```diff
@@ -39,25 +39,33 @@ internal static class OidcRetrieverWithCache
...
+                $"network for {authority}");
 
                 return configuration;
             }
             catch (Exception ex)
             {
-                requestContext.Logger.Error($"[OIDC Discovery] Failed to retrieve OpenID configuration from the OpenID endpoint {authority + Constants.WellKnownOpenIdConfigurationPath} due to {ex}");
+                requestContext.Logger.Error($"[OIDC Discovery] Failed to retrieve OpenID configuration from the OpenID endpoint " +
```
Comment (SimonCropp) [inline src/client/Microsoft.Identity.Client/Instance/Oidc/OidcRetrieverWithCache.cs:59]: remove the concat

File: src/client/Microsoft.Identity.Client/Instance/Oidc/OidcRetrieverWithCache.cs [hunk 9]
```diff
@@ -39,25 +39,33 @@ internal static class OidcRetrieverWithCache
...
+                    $"{authority + Constants.WellKnownOpenIdConfigurationPath} due to {ex}");
 
                 if (ex is MsalServiceException)
                     throw;
 
                 throw new MsalServiceException(
                     "oidc_failure",
-                    $"Failed to retrieve OIDC configuration from {authority + Constants.WellKnownOpenIdConfigurationPath}. See inner exception. ",
+                    $"Failed to retrieve OIDC configuration from {authority + Constants.WellKnownOpenIdConfigurationPath}. " +
```
Comment (SimonCropp) [inline src/client/Microsoft.Identity.Client/Instance/Oidc/OidcRetrieverWithCache.cs:67]: remove the concat

File: src/client/Microsoft.Identity.Client/Platforms/Features/SLC/KeyMaterialManager.cs [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+            {
+                if (s_bindingCertificate != null && !CertificateNeedsRotation(s_bindingCertificate))
+                {
+                    _logger.Verbose(() => "[Managed Identity] Another thread created the certificate while waiting for the lock.");
+                    _isInitialized = true;
+                    return s_bindingCertificate;
+                }
+
+                // The cached certificate needs to be rotated or does not exist
+                ECDsaCng cngkey = GetCngKey();
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Platforms/Features/SLC/KeyMaterialManager.cs:101]: Please add comments here on why you had to use the "cng" version? Afaik the "cng" versions are only supported on Windows? 

At least for RSA, the

**Comment by pmaytak** [comment]
@gladjohn Should this be set to draft; you're splitting it into a separate PR?

File: src/client/Microsoft.Identity.Client/ApiConfig/Executors/ManagedIdentityExecutor.cs [hunk 11]
```diff
@@ -35,19 +38,48 @@ public ManagedIdentityExecutor(IServiceBundle serviceBundle, ManagedIdentityAppl
         {
             var requestContext = CreateRequestContextAndLogVersionInfo(commonParameters.CorrelationId, cancellationToken);
 
+            if (_managedIdentityApplication.KeyMaterialManager.CryptoKeyType != CryptoKeyType.None)
+            {
+                // managed identity resource 
+                string miResource = managedIdentityParameters.Resource;
```
Comment (neha-bhargava) [inline src/client/Microsoft.Identity.Client/ApiConfig/Executors/ManagedIdentityExecutor.cs:44]: Extract this as a method in ScopeHelper.cs?

**Pull request closed without merging**